### DISCUSSION TITLE
How to force duckdb to work with system swaps

### DISCUSSION THREAD

**Discussion by WyattLiu** [duckdb/duckdb#7296] [Q&A]
I am working with large datasets that exceed my physical memory. How do I allow duckdb to leverage system swap when needed? Right now, regardless of in-memory or disk-backed. It always ends with "Out of Memory Error: failed to pin block of size X."

Please allow me to ask a few questions here:

1. Do we have any thoughts on enabling or disabling particular objects' swaps based on their nature to allow less memory requirement? I was reading some old threads where people complained about crashing. It looks like duckdb is implementing a use case by use case for the topic of disk offloading/swaps. 

2. Where should I start to find the existing offloading code?

I am wondering about the performance of swap to another machine's memory-based tmpfs via NFS. It should work.

**Comment by l1t1** [comment]
temp_directory of https://duckdb.org/docs/sql/configuration？

**Comment by l1t1** [comment]
```SQL
D:\duckdb>md tmp.tmp
D:\duckdb>duckdb071
v0.7.1 b00b93f0b1
Enter ".help" for usage hints.
Connected to a transient in-memory database.
Use ".open FILENAME" to reopen on a persistent database.
D SET memory_limit='10MB';
D create table t as select * from range(1e7::int)t(i);
Error: Out of Memory Error: could not allocate block of 262144 bytes (9984000/10000000 used)
Database is launched in in-memory mode and no temporary directory is specified.
Unused blocks cannot be offloaded to disk.

Launch the database with a persistent storage back-end
Or set PRAGMA temp_directory='/path/to/tmp.tmp'
D set temp_directory='tmp.tmp';
D create table t as select * from range(1e8::int)t(i);
Error: IO Error: Could not write file "tmp.tmp\duckdb_temp_storage-0.tmp" (error in WriteFile): ????????

D create table t as select * from range(1e7::int)t(i);
D select count(*) from t;
┌──────────────┐
│ count_star() │
│    int64     │
├──────────────┤
│     10000000 │
└──────────────┘
```

**Comment by l1t1** [comment]
seems wrong
D update t set i=i*2;
Error: Out of Memory Error: failed to allocate data of size 8388608

**Comment by l1t1** [comment]
```
D update t set i=i*2 where i<1e6;
Error: Out of Memory Error: failed to allocate data of size 8388608
D update t set i=i*2 where i<1e5;
D PRAGMA database_size;
┌───────────────┬───────────────┬────────────┬──────────────┬─────────────┬─────────────┬─────────
─┬──────────────┬──────────────┐
│ database_name │ database_size │ block_size │ total_blocks │ used_blocks │ free_blocks │ wal_size │ memory_usage │ memory_limit │
│    varchar    │    varchar    │   int64    │    int64     │    int64    │    int64    │ varchar  │   varchar    │   varchar    │
├───────────────┼───────────────┼────────────┼──────────────┼─────────────┼─────────────┼─────────
─┼──────────────┼──────────────┤
│ memory        │ 0 bytes       │          0 │            0 │           0 │           0 │ 0 bytes  │ 7.8MB        │ 10.0MB       │
└───────────────┴───────────────┴────────────┴──────────────┴─────────────┴─────────────┴─────────
─┴──────────────┴──────────────┘
D update t set i=i*2 where i<1e5;
D PRAGMA database_size;
┌───────────────┬───────────────┬────────────┬──────────────┬─────────────┬─────────────┬─────────
─┬──────────────┬──────────────┐
│ database_name │ database_size │ block_size │ total_blocks │ used_blocks │ free_blocks │ wal_size │ memory_usage │ memory_limit │
│    varchar    │    varchar    │   int64    │    int64     │    int64    │    int64    │ varchar  │   varchar    │   varchar    │
├───────────────┼───────────────┼────────────┼──────────────┼─────────────┼─────────────┼─────────
─┼──────────────┼──────────────┤
│ memory        │ 0 bytes       │          0 │            0 │           0 │           0 │ 0 bytes  │ 8.9MB        │ 10.0MB       │
└───────────────┴───────────────┴────────────┴──────────────┴─────────────┴─────────────┴─────────
─┴──────────────┴──────────────┘
```

**Comment by WyattLiu** [comment]
I think the documentation says the tmp file is created in the disk dir passed in as an argument, which greatly helps reduce the memory foot print. But the leverage of that tmp file seems to be class specific

**Reply by Alex-Monahan** [reply]
Howdy! When you say class specific, do you mean query specific? The group by / hash aggregate operator doesnot spill to disk yet (but it is on the roadmap!)

**Reply by WyattLiu** [reply]
Thank you for the road map updates.

**Comment by l1t1** [comment]
https://github.com/duckdb/duckdb/issues/58 says Off-loading to disk is only performed if a memory threshold is crossed

**Reply by Tishj** [reply]
Yea that's true, it's done as part of eviction, which only happens when we try to allocate a block that would cause us to exceed 
the memory limit.

**Reply by WyattLiu** [reply]
Do you know where the magic memory threshold is?

**Reply by Tishj** [reply]
https://duckdb.org/docs/sql/pragmas#memory_limit-threads

```
-- set the memory limit

PRAGMA memory_limit='1GB';
```

**Reply by WyattLiu** [reply]
Thank you, will try prob different limits. Hopefully, it would leverage more swaps.

**Comment by l1t1** [comment]
https://duckdb.org/2021/08/27/external-sorting.html says "can sort much more data than fits in main memory."
